File: parser_basic/read_rxs_log.m
function results = read_rxs_log(filename, maxCSINumber)
%read_rxs_log Parse PicoScenes generated .csi file using mex function RXSParser.
%   It invokes the mex function RXSParser to decode the data entries.
%   RXSParser will be automatically compiled, if the mex function does not exist.
%
    if nargin < 2
        maxCSINumber = intmax;
    end

    if exist('RXSParser', 'file') ~= 3
        compileRXSParser
    end
    
    if exist('RXSParser', 'file') ~= 3
        error('RXSParer compilation fails');
    end
    
    ticStart = tic;
    fp = fopen(filename, 'rb');
    if (fp < 0)
        error('Cannot open file %s', filename);
    end

    resultBatchSize = 10000;
    results = cell(resultBatchSize, 1);
    count = 1;

    % find maximal csi dimensions
    max_csi_dimensions = determine_csi_dimensions(fp, maxCSINumber)
    
    while ~feof(fp) && count <= maxCSINumber
        segmentLength = fread(fp, 1, 'uint32') + 4;
        if isempty(segmentLength)
            break;
        end
        fseek(fp, -4, 'cof');
        bytes=fread(fp, segmentLength,'uint8=>uint8');
        csi_entry = RXSParser(bytes);
        for i = 1:numel(csi_entry)
            csi_entry(i).MPDU = {csi_entry(i).MPDU};
        end
        if isempty(csi_entry) % in very rare case, the data is corrupted.
            continue;
        end

        % check that this frame's csi matrix dimensions are maximal
        % to always enable bundled parsing
        % replace with the previous frame if not, or skip if is first frame
        if ~isequal(size(csi_entry.CSI.CSI), max_csi_dimensions)
            if count ~= 1
                csi_entry = results{count - 1};
            else
                continue
            end
        end
        
        if count == numel(results)
            results = [results; cell(resultBatchSize, 1)];
        end

        results{count} = csi_entry;
        count = count + 1;
    end
    results(count:end)= [];

    disp([num2str(length(results)) ' PicoScenes frames are decoded in ' num2str(toc(ticStart)) ' seconds.']);
    fclose(fp);
    end

File: python_parser/read_rxs_log.py
import os
import struct
import time
from python_parser.compileRXSParser import compileRXSParser

def read_rxs_log(filename, maxCSINumber=None):
    if maxCSINumber is None:
        maxCSINumber = float('inf')

    if not os.path.exists('RXSParser.mex'):
        compileRXSParser()

    if not os.path.exists('RXSParser.mex'):
        raise Exception('RXSParser compilation fails')

    ticStart = time.time()
    with open(filename, 'rb') as fp:
        resultBatchSize = 10000
        results = [None] * resultBatchSize
        count = 0

        while True:
            segmentLength_bytes = fp.read(4)
            if not segmentLength_bytes:
                break
            segmentLength = struct.unpack('I', segmentLength_bytes)[0] + 4
            fp.seek(-4, os.SEEK_CUR)
            bytes = fp.read(segmentLength)
            csi_entry = RXSParser(bytes)
            for i in range(len(csi_entry)):
                csi_entry[i]['MPDU'] = [csi_entry[i]['MPDU']]
            if not csi_entry:
                continue

            # if ParserPreference.getPreference().skipBasebandSignals:
            #     csi_entry = [{k: v for k, v in entry.items() if k != 'BasebandSignals'} for entry in csi_entry]

            if count == len(results):
                results.extend([None] * resultBatchSize)
            results[count] = csi_entry
            count += 1

            if count > maxCSINumber:
                break

    results = results[:count]
    print(f'{len(results)} PicoScenes frames are decoded in {time.time() - ticStart} seconds.')

